wikia.php

```diff
@@ -48,41 +47,56 @@
 	// initialize skin if requested
 	$app->initSkin( (bool) $app->wg->Request->getVal( "skin", false ) );
 
-	$response = $app->sendExternalRequest( null, null, null );
-
-	// if cache policy wasn't explicitly set (e.g. WikiaResponse::setCacheValidity)
-	// then force no cache to reflect api.php default behavior
-	$cacheControl = $response->getHeader( 'Cache-Control' );
-
-	if ( empty( $cacheControl ) ) {
-		$response->setHeader( 'Cache-Control', 'private', true );
-
-		Wikia\Logger\WikiaLogger::instance()->info( 'wikia-php.caching-disabled', [
-			'controller' => $response->getControllerName(),
-			'method' => $response->getMethodName()
+	try {
+		$response = $app->sendExternalRequest( null, null, null );
+
+		// if cache policy wasn't explicitly set (e.g. WikiaResponse::setCacheValidity)
+		// then force no cache to reflect api.php default behavior
+		$cacheControl = $response->getHeader( 'Cache-Control' );
+
+		if ( empty( $cacheControl ) ) {
+			$response->setHeader( 'Cache-Control', 'private', true );
+
+			Wikia\Logger\WikiaLogger::instance()->info( 'wikia-php.caching-disabled', [
+				'controller' => $response->getControllerName(),
+				'method' => $response->getMethodName()
+			] );
+		}
+
+		// PLATFORM-1633: decrease the noise in reported transactions
+		$ex = $response->getException();
+
+		if ( $ex instanceof ControllerNotFoundException || $ex instanceof MethodNotFoundException ) {
+			Transaction::setAttribute( Transaction::PARAM_CONTROLLER, 'notFound' );
+			Transaction::setAttribute( Transaction::PARAM_METHOD, '' );
+
+			Wikia\Logger\WikiaLogger::instance()->info( 'wikia-php.not-found', [
+				'controller' => $response->getControllerName(),
+				'method' => $response->getMethodName()
+			] );
+		}
+
+		wfHandleCrossSiteAJAXdomain(); // PLATFORM-1719
+
+		$response->sendHeaders();
+		Hooks::run( 'NirvanaAfterRespond', [ $app, $response ] );
+
+		$response->render();
+	} catch ( WikiaHttpException $e ) {
+		http_response_code( $e->getCode() );
+		if ( $e->getCode() >= WikiaResponse::RESPONSE_CODE_INTERNAL_SERVER_ERROR ) {
+			Wikia\Logger\WikiaLogger::instance()->error( 'Unhandled API error', [
+				'status'=> $e->getCode(),
+				'exception' => $e
+			] );
+		}
```

Comment:
Should we render some kind of error body here?

Reply:
I'm not sure. I think API clients can specify the response format which may also involve a templating engine when generating a response. But I would not like to duplicate the code from `WikiaDispatcher.class.php` here.
We can return a simple JSON response though, just with a status code and error message. What do you think?

Reply:
A simple JSON response sounds good.

Reply:
That can work - alternatively, we could add another field to WikiaRequest that'd truly distinguish between PHP -> PHP nirvana calls and web requests, and update the dispatcher to take that into account when handling exceptions

Reply:
Added the JSON response. I think exception handling should be implemented regardless of additional flags - if something is broken, we should not silently ignore it and return 200 responses. Whether we need Yet Another Flag for PHP -> PHP nirvana calls - that is a good question.